src/rust/engine/process_execution/src/docker.rs

```diff
@@ -0,0 +1,360 @@
+
+    // DOCKER-TODO: Use a drop guard to remove the container regardless of success/failure.
+    // We cannot set `.host_config.auto_remove` in `config` above because containers that exit
+    // early will no longer exist and thus this code will not be able to attach and capture
+    // the output (which is the entire point of this executor ...).
+
+    self
+      .docker
+      .start_container::<String>(&container.id, None)
+      .await
+      .map_err(|err| {
+        format!(
+          "Failed to start Docker container `{}`: {:?}",
+          &container.id, err
+        )
+      })?;
+
+    log::trace!("started container {}", &container.id);
+
+    let attach_options = bollard::container::AttachContainerOptions::<String> {
+      stdout: Some(true),
+      stderr: Some(true),
+      logs: Some(true), // stream any output that was missed between the start_container call and now
+      stream: Some(true),
+      ..bollard::container::AttachContainerOptions::default()
+    };
+
+    let attach_result = self
+      .docker
+      .attach_container(&container.id, Some(attach_options))
+      .await
+      .map_err(|err| {
+        format!(
+          "Failed to attach to Docker container `{}`: {:?}",
+          &container.id, err
+        )
+      })?;
+
+    log::trace!("attached to container {}", &container.id);
+
+    let mut output_stream = attach_result.output.boxed();
+
+    let wait_options = bollard::container::WaitContainerOptions {
+      condition: "not-running",
+    };
+    let mut wait_stream = self
+      .docker
+      .wait_container(&container.id, Some(wait_options))
+      .boxed();
+
+    let container_id = container.id.to_owned();
+    let result_stream = async_stream::stream! {
+      let mut run_stream = true;
+      while run_stream {
+        tokio::select! {
+          Some(output_msg) = output_stream.next() => {
+            match output_msg {
+              Ok(LogOutput::StdOut { message }) => {
+                log::trace!("container {} wrote {} bytes to stdout", &container_id, message.len());
+                yield Ok(ChildOutput::Stdout(message));
+              }
+              Ok(LogOutput::StdErr { message }) => {
+                log::trace!("container {} wrote {} bytes to stderr", &container_id, message.len());
+                yield Ok(ChildOutput::Stderr(message));
+              }
+              _ => (),
+            }
+          }
+          Some(wait_msg) = wait_stream.next() => {
+            log::trace!("wait_container stream ({}): {:?}", &container_id, wait_msg);
+            match wait_msg {
+              Ok(r) => {
+                // DOCKER-TODO: How does Docker distinguish signal versus exit code? Improve
+                // `ChildResults` to better support exit code vs signal vs error message?
+                let status_code = r.status_code;
+                yield Ok(ChildOutput::Exit(ExitCode(status_code as i32)));
+                run_stream = false;
```

Comment:
Is there a particular reason to use `while run_stream`, rather than `loop` + `break`?

Reply:
Done.